<a href="https://colab.research.google.com/github/veeransr/ML-and-Deep-learning-practice-with-python/blob/master/imp_other_assignment_Reg_kerass_optimization_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import required libraries

In [0]:
import pandas as pd
import numpy as np

Load source data

In [0]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

Check for missing data 

In [0]:
concrete_data.isna().any()

Cement                False
Blast Furnace Slag    False
Fly Ash               False
Water                 False
Superplasticizer      False
Coarse Aggregate      False
Fine Aggregate        False
Age                   False
Strength              False
dtype: bool

Separate target and predictors (pop in the strength column from concrete_data)

In [0]:
target = concrete_data.pop('Strength')
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Check that concrete_data does not contain strength information

In [0]:
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

Build the model with 
* Three hidden layers, each of 10 nodes and a ReLU activation function
* Use the adam optimizer and the mean squared error as the loss function

In [0]:
from keras.models import Sequential
from keras.layers import Dense

def build_model(input_shape):
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(input_shape,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    model.compile(loss='mean_squared_error',
                 optimizer='adam')
    model.summary()
    return model

## 1. Randomly split the data into a training and test sets by holding 30% of the data for testing.

In [0]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(concrete_data, target, test_size=0.30)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(721, 8) (309, 8) (721,) (309,)


Normalize the train data

In [0]:
train_means = X_train.mean()
train_stds = X_train.std()
X_train_norm = (X_train - train_means) / train_stds
X_test_norm = (X_test - train_means) / train_stds

In [0]:
X_train_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
329 -0.347363           -0.839882  1.126289 -1.785793          0.971787   
130  0.374259            2.412951 -0.839100  0.105221          0.686106   
107  0.374259            2.412951 -0.839100  0.105221          0.686106   
416 -0.659847           -0.839882  1.897677 -1.043395          0.921373   
271 -0.488121           -0.839882  1.071302 -0.352357          0.081134   

     Coarse Aggregate  Fine Aggregate       Age  
329          1.441764        0.354116 -0.692536  
130         -0.399263       -1.405852 -0.276853  
107         -0.399263       -1.405852 -0.626026  
416          0.889839       -0.007863 -0.692536  
271          1.053372        0.074518 -0.276853

In [0]:
X_test_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
273 -0.488121           -0.839882  1.071302 -0.352357          0.081134   
984 -1.210682           -0.839882  1.404366  0.548793          0.484448   
233 -0.657970            0.288488 -0.454192  0.007169          0.114743   
172  1.324847            0.382806 -0.839100 -1.309537          1.728002   
244 -0.429003           -0.839882  0.639262  0.240627          0.131548   

     Coarse Aggregate  Fine Aggregate       Age  
273          1.053372        0.074518  0.920314  
984         -1.236093        1.187917 -0.276853  
233          1.173467        0.160644  0.920314  
172         -1.556771        1.430069  0.770668  
244         -0.307275        0.929539 -0.692536

Build model

In [0]:
concrete_model = build_model(X_train_norm.shape[1])

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 10)                90        
_________________________________________________________________
dense_2 (Dense)              (None, 10)                110       
_________________________________________________________________
dense_3 (Dense)              (None, 10)                110       
_________________________________________________________________
dense_4 (Dense)              (None, 1)                 11        
Total params: 321
Trainable params: 321
Non-trainable params: 0
_________________________________________________________________


## 2. Train the model on the training data using 50 epochs

In [0]:
concrete_model.fit(X_train_norm, y_train, epochs=50, verbose=0)

## 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength

Use the model to predict strength

In [0]:
y_hat = concrete_model.predict(X_test_norm)
# print(y_hat)

In [0]:
from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_test, y_hat)
mse

146.29454495431634

## 4. Repeat steps 1 - 3, 50 times

In [0]:
mse_list = list()

In [0]:
for r in range(0, 50):
    # 1
    X_train, X_test, y_train, y_test = train_test_split(concrete_data, target, test_size=0.30)
    train_means = X_train.mean()
    train_stds = X_train.std()
    X_train_norm = (X_train - train_means) / train_stds
    X_test_norm = (X_test - train_means) / train_stds
    # 2 
    concrete_model.fit(X_train_norm, y_train, epochs=50, verbose=0)
    # 3
    y_hat = concrete_model.predict(X_test_norm)
    mse = mean_squared_error(y_test, y_hat)
    mse_list.append(mse)

## 5. Report the mean and the standard deviation of the mean squared errors.

In [0]:
print("Mean of the 50 runs = {} /// Standard deviation = {}".format(sum(mse_list)/len(mse_list), np.std(mse_list)))

Mean of the 50 runs = 38.621805991374515 /// Standard deviation = 22.676398014814378


Compare to Step B the mean squared error is better. The mean has improved but the standard deviation just a little. Increasing hidden layers in the models enables to have in average better predictions.
* Results of step B mean = 41.950 /// std = 23.994
* Results of step D mean = 38.622 /// std = 22.676